### ISSUE TITLE
Performance / Memory uasage concerns for cache over large table. 50m+

### ISSUE THREAD

**Issue by akua28** [zzzprojects/EntityFramework-Plus#160]
HI All,

I just found out this liberary and it looks promising for the project i'm working on.
But, before i start using it i would like to ask some opinions.

My project is a multitanent website which will have a table with 50m+ rows with annual growth rate over 10m+. I'm a bit concerned with the how much memory is needed for the cache.

Atm I can't find any doc regarding how cache is handled when memory is intense. 
Can anyone share some experience or insight?

Kind Regards

**Comment by JonathanMagnan** [comment]
Hello @akua28 ,

I cannot provide my opinion since mine is highly biased.

The amount of memory that will be required highly depend on what will be cached.

If you cache some data like countries && states, you will not have any issue since only thousands of entities will be cached. However, if you start to cache millions of data, you will for sure face some kind of memory issue soon or later.

I believe to get a better answer/feedback, you should explain more the kind of scenario you would like to implement.

Best Regards,

Jonathan

**Comment by akua28** [comment]
Hi @JonathanMagnan,

The scenario we are facing is like this:
We run a query which will return about 2m+ rows. We then need to reuse that query result for paging or calculation on all the rows based on web requests.
Following the project's API, the 2m+ rows will be cached by `Context.Table.Query.FromCache()`.
These are my concerns / want to knows:
- If server reach memory limit, will some of the cache be droped (FIFO queue? Or completely)
- Is the cache implemented using an plain object collection? Or an in memory index (or table) that can be queried over?
- In the example the syntax for query is fairly simple, `Context.Table.Where(a => a.IsActive).FromCache()`.
Can it cache more complex predicts? E.g. at least 5 wheres.
Can it cache join queries?

Kind Regards
Anakin

**Comment by JonathanMagnan** [comment]
Hello Anakin,

Our query cache implementation simply caches the result in a Memory Cache ([Code Source](https://github.com/zzzprojects/EntityFramework-Plus/blob/master/src/Z.EntityFramework.Plus.EF6/QueryCache/QueryCacheManager.cs#L43)) as you can easily do the same on your side. We just make it a little bit easier by providing extension methods to IQueryable interface.

It's not a smart cache that will cache the entities only once as you may find with some other cache library that exists.

So yes, it can cache join queries, complex predicts (entities will be duplicated) and has some memory limit since that's your job to configure the cache configuration: See [MemoryCache](https://msdn.microsoft.com/en-us/library/system.runtime.caching.memorycache(v=vs.110).aspx). After all, our library simply cache the result using the LINQ queries as key.

Typically, it's recommended to perform this kind of operation on the database side but perhaps it impossible because of some custom calculation coming from the web requests.

So in short, you will find the answer by looking how to work the MemoryCache and if you believe it fit or not your scenario.

Let me know if that helped you a little bit or you need more information

Best Regards,

Jonathan

**Comment by JonathanMagnan** [comment]
Closing comment: No confirmation, feel free to re-open if you still have the issue.

**Comment by akua28** [comment]
Thanks for the additional info.
